packages/salesforcedx-vscode-soql/src/client/codeCompletion.ts

```diff
@@ -0,0 +1,111 @@
+ */
+
+import { CompletionItemKind } from 'vscode';
+import ProtocolCompletionItem from 'vscode-languageclient/lib/protocolCompletionItem';
+import { retrieveSObject, retrieveSObjects } from '../sfdx';
+
+import { Middleware } from 'vscode-languageclient';
+import { FieldType } from 'jsforce';
+
+export const middleware: Middleware = {
+  // The SOQL LSP server may include special completion items as "placeholders" for
+  // the client to expand with information from the users' default Salesforce Org.
+  // We do that here as middleware, transforming the server response before passing
+  // it up to VSCode.
+
+  provideCompletionItem: async (document, position, context, token, next) => {
+    const items = (await next(
+      document,
+      position,
+      context,
+      token
+    )) as ProtocolCompletionItem[];
+
+    return expandPlaceholders(items);
+  }
+};
+
+const expandableItemPattern = /__([A-Z_]+)(:(\w+(,\w+)*))?/;
+async function expandPlaceholders(
+  items: ProtocolCompletionItem[]
+): Promise<ProtocolCompletionItem[]> {
+  const expandedItems = [...items];
+
+  items.forEach(async (item, index) => {
+    const m = item.label.match(expandableItemPattern);
+    if (m) {
+      const command = m[1];
+      const args = m[3] ? m[3].split(',') : [];
+
+      const handler = expandFunctions[command];
+      if (!handler) {
+        console.log(`Unknown SOQL completion command ${command}!`);
+      }
+
+      expandedItems.splice(index, 1, ...(await handler(args)));
+    }
+  });
+
+  return expandedItems;
+}
+
+const expandFunctions: {
+  [key: string]: (args: string[]) => Promise<ProtocolCompletionItem[]>;
+} = {
+  SOBJECTS_PLACEHOLDER: async (
+    args: string[]
+  ): Promise<ProtocolCompletionItem[]> => {
+    try {
+      const sobjectItems = (await retrieveSObjects()).map(objName => {
+        const item = new ProtocolCompletionItem(objName);
+        item.kind = CompletionItemKind.Class;
+        return item;
+      });
+      return sobjectItems;
+    } catch (metadataErrors) {
+      return [];
+    }
+  },
+
+  SOBJECT_FIELDS_PLACEHOLDER: async (
+    args: string[]
+  ): Promise<ProtocolCompletionItem[]> => {
+    try {
+      const objMetadata = await retrieveSObject(args[0]);
+      const sobjectFields = objMetadata.fields.reduce((fieldItems, field) => {
+        fieldItems.push(newFieldCompletionItem(field.name, field.name));
+
+        if (field.relationshipName) {
+          fieldItems.push(
+            newFieldCompletionItem(
+              `${field.relationshipName} \(${field.referenceTo}\)`,
+              field.relationshipName + '.',
+              CompletionItemKind.Class
+            )
+          );
+        }
+
+        return fieldItems;
+      }, [] as ProtocolCompletionItem[]);
+      return sobjectFields;
+    } catch (metadataError) {
+      return [];
+    }
+  }
+};
+
+function newFieldCompletionItem(
```

Comment:
I think these changes should be part of your completion PR.

Reply:
It's just code moved from client.ts into a separate file. See commit dc95f508c5f116b5b39e3d08251b62ab4a8a33cb